include/libtorrent/alert_manager.hpp

```diff
@@ -160,16 +223,141 @@ namespace libtorrent {
 		alert_manager(alert_manager const&);
 		alert_manager& operator=(alert_manager const&);
 
-		void maybe_notify(alert* a, mutex::scoped_lock& lock);
+		void init_thread_specific_storage()
+		{
+			thread_state ts;
+			ts.youngest_generation = new boost::atomic<int>(0);
+			ts.oldest_generation = 0;
+			ts.n_generations = 0;
+
+			m_generation.reset(ts.youngest_generation);
+
+			for (int i = 0; i < TORRENT_ALERT_MANAGER_N_GENERATIONS; i++)
+			{
+				ts.allocators[i] = new aux::stack_allocator();
+				m_allocations[i].reset(ts.allocators[i]);
+			}
+
+			// register storage globally
+			mutex::scoped_lock lock(m_mutex);
+			m_threads.push_back(ts);
+		}
+
+		// tries to emplace a pre-constructed alert. Returns false if the
+		// queue fills up before the alert can be posted true otherwise.
+		bool do_emplace_alert(alert * const a, const int priority, const int& gen, bool& aborted)
+		{
+			int current, next;
+			alert* expected;
+
+			TORRENT_ASSERT(a != NULL);
+
+			// reference the ring buffer to ensure it's not resized
+			m_buffer_refs++;
+
+			// if a resize has been requested we must lock
+			// the mutex before taking a reference
+			if (m_queue_limit_requested.load(boost::memory_order_relaxed) != -1) {
```

Comment:
It is guaranteed to happen after `m_buffer_refs++` because that one is atomic. But nothing depends on it happening immediately though, just like `m_alert_mask` where it doesn't matter if the change from 0 to 1 is not seen immediately.

Reply:
I forgot to mention why. It's because `m_buffer_refs` is atomic. I should use `m_buffer_refs.fetch_add(1)` instead of `m_buffer_refs++` to make it clear to the reader that this is a synchronized operation. Actually that relaxed load cannot be moved at all because it's sandwished between synchronized operations. But if we change `const int size_limit = m_queue_size_limit` to `const int size_limit = m_queue_size_limit.load(boost::memory_order_consume)` it should preserve the same behaviour but may allow some optimizations. For example, on a 64-bit cpu `m_queue_limit_requested` can be loaded by the same instruction as `m_buffer_size`, because we're not asking the compiler to ensure that `m_buffer_size` is loaded after `m_queue_limit_requested`. We would have move them so they're sequential and make sure they're properly aligned so this is just an example.